specs/sharding/beacon-chain.md

```diff
@@ -500,45 +575,67 @@ def process_operations(state: BeaconState, body: BeaconBlockBody) -> None:
     for_ops(body.attester_slashings, process_attester_slashing)
     # New shard proposer slashing processing
     for_ops(body.shard_proposer_slashings, process_shard_proposer_slashing)
-    # Limit is dynamic based on active shard count
+
+    # Limit is dynamic: based on active shard count
     assert len(body.shard_headers) <= MAX_SHARD_HEADERS_PER_SHARD * get_active_shard_count(state, get_current_epoch(state))
     for_ops(body.shard_headers, process_shard_header)
+
     # New attestation processing
     for_ops(body.attestations, process_attestation)
     for_ops(body.deposits, process_deposit)
     for_ops(body.voluntary_exits, process_voluntary_exit)
+
+    # TODO: to avoid parallel shards racing, and avoid inclusion-order problems,
+    #  update the fee price per slot, instead of per header.
+    # state.shard_sample_price = compute_updated_sample_price(state.shard_sample_price, ?, shard_count)
 ```
 
 ##### Extended Attestation processing
 
 ```python
 def process_attestation(state: BeaconState, attestation: Attestation) -> None:
     altair.process_attestation(state, attestation)
-    update_pending_shard_work(state, attestation)
+    process_attested_shard_work(state, attestation)
 ```
 
 ```python
-def update_pending_shard_work(state: BeaconState, attestation: Attestation) -> None:
+def process_attested_shard_work(state: BeaconState, attestation: Attestation) -> None:
     attestation_shard = compute_shard_from_committee_index(
         state,
         attestation.data.slot,
         attestation.data.index,
     )
+    full_committee = get_beacon_committee(state, attestation.data.slot, attestation.data.index)
+
     buffer_index = attestation.data.slot % SHARD_STATE_MEMORY_SLOTS
     committee_work = state.shard_buffer[buffer_index][attestation_shard]
 
     # Skip attestation vote accounting if the header is not pending
     if committee_work.status.selector != SHARD_WORK_PENDING:
-        # TODO In Altair: set participation bit flag, if attestation matches winning header.
+        # If the data was already confirmed, check if this matches, to apply the flag to the attesters.
+        if committee_work.status.selector == SHARD_WORK_CONFIRMED:
+            attested: AttestedDataCommitment = committee_work.status.value
+            if attested.root == attestation.data.shard_blob_root:
+                batch_apply_participation_flag(state, attestation.aggregation_bits,
+                                               attestation.data.target.epoch,
+                                               full_committee, TIMELY_SHARD_FLAG_INDEX)
         return
 
     current_headers: Sequence[PendingShardHeader] = committee_work.status.value
 
     # Find the corresponding header, abort if it cannot be found
-    header_index = [header.root for header in current_headers].index(attestation.data.shard_header_root)
+    header_index = len(current_headers)
```

Comment:
`index()` would abort with an exception if the data was not found, which is the kind of silent edge case I would like to avoid (and it needs to return without exception here). And instead of creating temporary lists, we can just implement the behavior directly (readable and fast). Might just be me thinking in Go style too much though.

Reply:
```python
if not shard_blob_root in current_header_roots:
    return

header_index = current_header_roots.index(shard_blob_root)
```

This has a check above to ensure the item is certainly in the list when call to `index()`.

This does feel a bit off wrt previous style, but we can wait until the great (inevitable) cleanups come through